Detectors/EMCAL/reconstruction/src/CaloRawFitterGamma2.cxx

```diff
@@ -0,0 +1,172 @@
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file CaloRawFitterGamma2.cxx
+/// \author Martin Poghosyan (Martin.Poghosyan@cern.ch)
+
+#include "FairLogger.h"
+#include <random>
+
+// ROOT sytem
+#include "TMath.h"
+
+#include "EMCALReconstruction/Bunch.h"
+#include "EMCALReconstruction/CaloFitResults.h"
+#include "DataFormatsEMCAL/Constants.h"
+
+#include "EMCALReconstruction/CaloRawFitterGamma2.h"
+
+using namespace o2::emcal;
+
+CaloRawFitterGamma2::CaloRawFitterGamma2() : CaloRawFitter("Chi Square ( Gamma2 )", "Gamma2")
+{
+  mAlgo = FitAlgorithm::Gamma2;
+}
+
+CaloFitResults CaloRawFitterGamma2::evaluate(const std::vector<Bunch>& bunchlist,
+                                             std::optional<unsigned int> altrocfg1, std::optional<unsigned int> altrocfg2)
+{
+  float time = 0;
+  float amp = 0;
+  float chi2 = 0;
+  int ndf = 0;
+  bool fitDone = false;
+
+  auto [nsamples, bunchIndex, ampEstimate,
+        maxADC, timeEstimate, pedEstimate, first, last] = preFitEvaluateSamples(bunchlist, altrocfg1, altrocfg2, mAmpCut);
+
+  if (ampEstimate >= mAmpCut) {
+    time = timeEstimate;
+    int timebinOffset = bunchlist.at(bunchIndex).getStartTime() - (bunchlist.at(bunchIndex).getBunchLength() - 1);
+    amp = ampEstimate;
+
+    if (nsamples > 2 && maxADC < constants::OVERFLOWCUT) {
+      doParabolaFit(timeEstimate - 1, amp, time);
+      mNiter = 0;
+      fitDone = doFit_1peak(first, nsamples, amp, time, chi2);
+
+      if (!fitDone) {
+        amp = ampEstimate;
+        time = timeEstimate;
+        chi2 = 1.e9;
+      }
+
+      time += timebinOffset;
+      timeEstimate += timebinOffset;
+      ndf = nsamples - 2;
+    }
+  }
+
+  if (fitDone) {
+    float ampAsymm = (amp - ampEstimate) / (amp + ampEstimate);
+    float timeDiff = time - timeEstimate;
+
+    if ((TMath::Abs(ampAsymm) > 0.1) || (TMath::Abs(timeDiff) > 2)) {
+      amp = ampEstimate;
+      time = timeEstimate;
+      fitDone = false;
+    }
+  }
+  if (amp >= mAmpCut) {
+    if (!fitDone) {
+      std::default_random_engine generator;
+      std::uniform_real_distribution<float> distribution(0.0, 1.0);
+      amp += (0.5 - distribution(generator));
+    }
+    time = time * constants::EMCAL_TIMESAMPLE;
+    time -= mL1Phase;
+
+    return CaloFitResults(maxADC, pedEstimate, mAlgo, amp, time, (int)time, chi2, ndf);
+  }
+  return CaloFitResults(-1, -1);
+}
+
+bool CaloRawFitterGamma2::doFit_1peak(int ifirst, int nS, float& A0, float& t0, float& chi2)
+{
+  // fit using gamma-2 function 	(ORDER =2 assumed)
+
+  if (nS < 3)
+    return false;
+  if (mNiter > mNiterationsMax)
+    return false;
+
+  double expi, ti, g_i, g_1i, g_2i, g1_1, g2_i, gp_i, g2p_i, q1_i, q2_i, D, dA, dt, delta;
+  double c11 = 0;
+  double c12 = 0;
+  double c21 = 0;
+  double c22 = 0;
+  double d1 = 0;
+  double d2 = 0;
```

Comment:
Variable names should be meaningful.

Reply:
I will ask Martin for help concerning these parameters.